File: 322.coin-change.go
func coinChange(coins []int, amount int) int {
    amounts := make([]int, amount+1)
    for i := 1; i < len(amounts); i++ {
        mn := -1
        for j := range coins {
            minused := i - coins[j]
            if minused == 0 {
                mn = 1
                break
            } else if minused > 0 && amounts[minused] > 0 && (mn < 0 || (mn > 0  && amounts[minused] + 1 < mn)) {
                    mn = amounts[minused] + 1
                } 
        }
        amounts[i] = mn
    }
    return amounts[amount]
}

File: 322.coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        mem = [-1] * (amount + 1)
        for i in range(len(mem)):
            if i <= 0:
                i = -1
            elif i in coins:
                mem[i] = 1
            else:
                for j in coins:
                    if i - j >= 0 and mem[i - j] > 0:
                        mem[i] = min(mem[i] if mem[i] > 0 else 10001, mem[i - j] + 1)
        # print(mem)
        return mem[-1]
